File: js/Totobot.js
// Totobot.js

(function (ext) {
	var device = null;

	var devices = {
		VERSION: 0,
		JOYSTICK: 5,
		MOTOR: 10,
		DIGITAL: 30,
		ANALOG: 31,
		PWM: 32,
		ULTRASONIC_ARDUINO: 36,
		PULSEIN: 37,
		LEDMATRIX: 41,
		TOTO_MOTOR: 90,
		TOTO_MOTOR_RANGE: 91,
		TOTO_JOYSTICK: 92,
		TOTO_EYE_EFFECT: 93
	};

	var directions = {
		"move forward": 1,
		"move backward": 2,
		"turn left": 3,
		"turn right": 4
	}

	var motorPorts = {
		"left motor": 1,
		"right motor": 2
	}

	var eyes = {
		"right eye": 0,
		"left eye": 1
	}

	var effects = { "none": 0, "snow": 2, "matrix": 7 }

	ext.resetAll = function () {
	};

	ext.runArduino = function() {
		responseValue();
	};

	ext.moveForward = function (speed, duration) {
		runPackage(devices.TOTO_JOYSTICK, short2array(speed * 255 / 100), short2array(speed * 255 / 100));
		sleep(duration * 1000);
		runPackage(devices.TOTO_JOYSTICK, short2array(0), short2array(0));
	};

	ext.moveBackward = function (speed, duration) {
		runPackage(devices.TOTO_JOYSTICK, short2array(-speed * 255 / 100), short2array(-speed * 255 / 100));
		sleep(duration * 1000);
		runPackage(devices.TOTO_JOYSTICK, short2array(0), short2array(0));
	};

	ext.turnLeft = function (speed, duration) {
		runPackage(devices.TOTO_JOYSTICK, short2array(-speed * 255 / 100), short2array(speed * 255 / 100));
		sleep(duration * 1000);
		runPackage(devices.TOTO_JOYSTICK, short2array(0), short2array(0));
	};

	ext.turnRight = function (speed, duration) {
		runPackage(devices.TOTO_JOYSTICK, short2array(speed * 255 / 100), short2array(-speed * 255 / 100));
		sleep(duration * 1000);
		runPackage(devices.TOTO_JOYSTICK, short2array(0), short2array(0));
	};

	ext.move = function (direction, speed) {
		if (typeof direction == "string") {
			direction = directions[direction];
		}
		var leftSpeed = direction == 1 || direction == 4 ? speed : -speed;
		var rightSpeed = direction == 1 || direction == 3 ? speed : -speed;
		runPackage(devices.TOTO_JOYSTICK, short2array(leftSpeed * 255 / 100), short2array(rightSpeed * 255 / 100));
	};

	ext.runMotors = function (leftSpeed, rightSpeed) {
		runPackage(devices.TOTO_JOYSTICK, short2array(leftSpeed * 255 / 100), short2array(rightSpeed * 255 / 100));
    };

	ext.stopMotors = function () {
		runPackage(devices.TOTO_JOYSTICK, short2array(0), short2array(0));
	};

	ext.setMotorRange = function (motorPort, min, max) {
		if (typeof motorPort == "string") {
			motorPort = motorPorts[motorPort];
		}
		runPackage(devices.TOTO_MOTOR_RANGE, motorPort, min, max);
	};

	ext.setEyeEffect = function (eye, effect) {
		if (typeof eye == "string") {
			eye = eyes[eye];
		}
		if (typeof effect == "string") {
			effect = effects[effect];
		}
		runPackage(devices.TOTO_EYE_EFFECT, eye, short2array(effect));
	};

	/**
	 * @deprecated
	 * @param {number} delay
	 */
	function sleep(delay) {
		var start = new Date().getTime();
		while (new Date().getTime() < start + delay);
	}

	function sendPackage(argList, type) {
		var bytes = [0xff, 0x55, 0, 0, type];
		for (var i = 0; i < argList.length; ++i) {
			var val = argList[i];
			if (val.constructor == "[class Array]") {
				bytes = bytes.concat(val);
			} else {
				bytes.push(val);
			}
		}
		bytes[2] = bytes.length - 3;
		device.send(bytes);
	}
	
	function runPackage() {
		sendPackage(arguments, 2);
	}
	function getPackage() {
		var nextID = arguments[0];
		Array.prototype.shift.call(arguments);
		sendPackage(arguments, 1);
	}

	var inputArray = [];
	var _isParseStart = false;
	var _isParseStartIndex = 0;
	function processData(bytes) {
		var len = bytes.length;
		if (_rxBuf.length > 30) {
			_rxBuf = [];
		}
		for (var index = 0; index < bytes.length; index++) {
			var c = bytes[index];
			_rxBuf.push(c);
			if (_rxBuf.length >= 2) {
				if (_rxBuf[_rxBuf.length - 1] == 0x55 && _rxBuf[_rxBuf.length - 2] == 0xff) {
					_isParseStart = true;
					_isParseStartIndex = _rxBuf.length - 2;
				}
				if (_rxBuf[_rxBuf.length - 1] == 0xa && _rxBuf[_rxBuf.length - 2] == 0xd && _isParseStart) {
					_isParseStart = false;

					var position = _isParseStartIndex + 2;
					var extId = _rxBuf[position];
					position++;
					var type = _rxBuf[position];
					position++;
					//1 byte 2 float 3 short 4 len+string 5 double
					var value;
					switch (type) {
						case 1: {
							value = _rxBuf[position];
							position++;
						}
							break;
						case 2: {
							value = readFloat(_rxBuf, position);
							position += 4;
							if (value < -255 || value > 1023) {
								value = 0;
							}
						}
							break;
						case 3: {
							value = readInt(_rxBuf, position, 2);
							position += 2;
						}
							break;
						case 4: {
							var l = _rxBuf[position];
							position++;
							value = readString(_rxBuf, position, l);
						}
							break;
						case 5: {
							value = readDouble(_rxBuf, position);
							position += 4;
						}
							break;
						case 6:
							value = readInt(_rxBuf, position, 4);
							position += 4;
							break;
					}
					if (type <= 6) {
						responseValue(extId, value);
					} else {
						responseValue();
					}
					_rxBuf = [];
				}
			}
		}
	}
	function readFloat(arr, position) {
		var f = [arr[position], arr[position + 1], arr[position + 2], arr[position + 3]];
		return parseFloat(f);
	}
	function readInt(arr, position, count) {
		var result = 0;
		for (var i = 0; i < count; ++i) {
			result |= arr[position + i] << (i << 3);
		}
		return result;
	}
	function readDouble(arr, position) {
		return readFloat(arr, position);
	}
	function readString(arr, position, len) {
		var value = "";
		for (var ii = 0; ii < len; ii++) {
			value += String.fromCharCode(_rxBuf[ii + position]);
		}
		return value;
	}
	function appendBuffer(buffer1, buffer2) {
		return buffer1.concat(buffer2);
	}

	// Extension API interactions
	var potentialDevices = [];
	ext._deviceConnected = function (dev) {
		potentialDevices.push(dev);

		if (!device) {
			tryNextDevice();
		}
	}

	function tryNextDevice() {
		// If potentialDevices is empty, device will be undefined.
		// That will get us back here next time a device is connected.
		device = potentialDevices.shift();
		if (device) {
			device.open({ stopBits: 0, bitRate: 115200, ctsFlowControl: 0 }, deviceOpened);
		}
	}

	var watchdog = null;
	function deviceOpened(dev) {
		if (!dev) {
			// Opening the port failed.
			tryNextDevice();
			return;
		}
		device.set_receive_handler('Totobot', processData);
	};

	ext._deviceRemoved = function (dev) {
		if (device != dev) return;
		device = null;
	};

	ext._shutdown = function () {
		if (device) device.close();
		device = null;
	};

	ext._getStatus = function () {
		if (!device) return { status: 1, msg: 'Totobot disconnected' };
		if (watchdog) return { status: 1, msg: 'Probing for Totobot' };
		return { status: 2, msg: 'Totobot connected' };
	}

	var descriptor = {};
	ScratchExtensions.register('Totobot', descriptor, ext, { type: 'serial' });
})({});

File: src/Totobot.cpp
/*
	Totobot.cpp - library for Totobot - implementation
	Copyright (c) 2019 Vitaliy.  All right reserved.
	Used effects from https://github.com/AlexGyver/GyverMatrixBT/
*/

#include "Totobot.h"
#include "TotobotFace.h"
#include "TotobotOnlineListener.h"

#include <avr/eeprom.h>
#include "AFMotor.h"
#include "TimerOne.h"

Totobot totobot;

AF_DCMotor motors[] = {1, 2, 3, 4};
byte motorsMin[] = {0, 0, 0, 0};
byte motorsMax[] = {255, 255, 255, 255};

void Totobot::init() {
	reset();

	TotobotFace::init();

	Timer1.initialize(100000);
	Timer1.attachInterrupt(timer);

	TotobotOnlineListener::init();

	// read/init EEPROM
	if (eeprom_read_byte((byte *)EEPROM_BEGIN) == EEPROM_MAGIC_NUMBER &&
	    eeprom_read_byte((byte *)EEPROM_BEGIN + 1) == EEPROM_DATA_SIZE) {
		for (int i = 0; i < (sizeof motors / sizeof motors[0]); ++i) {
			motorsMin[i] = eeprom_read_byte((byte *)EEPROM_MOTORS_BEGIN + i * 2);
			motorsMax[i] = eeprom_read_byte((byte *)EEPROM_MOTORS_BEGIN + i * 2 + 1);
		}
	} else {
		eeprom_update_byte((byte *)EEPROM_BEGIN, EEPROM_MAGIC_NUMBER);
		eeprom_update_byte((byte *)EEPROM_BEGIN + 1, EEPROM_DATA_SIZE);
		for (int i = 0; i < (sizeof motors / sizeof motors[0]); ++i) {
			eeprom_update_byte((byte *)EEPROM_MOTORS_BEGIN + i * 2, motorsMin[i]);
			eeprom_update_byte((byte *)EEPROM_MOTORS_BEGIN + i * 2 + 1, motorsMax[i]);
		}
	}

	pinMode(LED_PIN, OUTPUT);
	digitalWrite(LED_PIN, HIGH);
	delay(300);
	digitalWrite(LED_PIN, LOW);
}

void Totobot::loop() {
	TotobotOnlineListener::loop();
}

void Totobot::timer() {
	TotobotFace::loop();
}

void Totobot::reset() {
	for (int i = 0; i < (sizeof motors / sizeof motors[0]); ++i)
		motors[i].run(RELEASE);
}

void Totobot::setMotorRange(byte number, byte min, byte max) {
	int index = (number - 1) % (sizeof motors / sizeof motors[0]);
	motorsMin[index] = min;
	motorsMax[index] = max;
	eeprom_update_byte((byte *)EEPROM_MOTORS_BEGIN + index * 2, min);
	eeprom_update_byte((byte *)EEPROM_MOTORS_BEGIN + index * 2 + 1, max);
}

void Totobot::runMotor(byte number, short speed) {
	int index = (number - 1) % (sizeof motors / sizeof motors[0]);
	if (index == LEFT_MOTOR - 1)
		speed = -speed; // turn of the left motor by analogy with mBot for M1
	byte absSpeed = speed >= 0 ? speed : -speed;
	absSpeed = map(absSpeed, 0, 255, motorsMin[index], motorsMax[index]);
	motors[index].setSpeed(absSpeed);
	motors[index].run(speed > 0 ? FORWARD : speed < 0 ? BACKWARD : RELEASE);
}

void Totobot::move(byte direction, byte speed) {
	switch (direction) {
	case MOVE_FORWARD: {
		runMotor(LEFT_MOTOR, speed);
		runMotor(RIGHT_MOTOR, speed);
	} break;
	case MOVE_BACKWARD: {
		runMotor(LEFT_MOTOR, -speed);
		runMotor(RIGHT_MOTOR, -speed);
	} break;
	case TURN_LEFT: {
		runMotor(LEFT_MOTOR, -speed);
		runMotor(RIGHT_MOTOR, speed);
	} break;
	case TURN_RIGHT: {
		runMotor(LEFT_MOTOR, speed);
		runMotor(RIGHT_MOTOR, -speed);
	} break;
	default: {
		runMotor(LEFT_MOTOR, 0);
		runMotor(RIGHT_MOTOR, 0);
	}
	}
}
void Totobot::ledShowImage(byte *bytes, byte size, byte x, byte y) {
	TotobotFace::showImage(bytes, size, x, y);
}

void Totobot::ledShowImage(byte b0, byte b1, byte b2, byte b3, byte b4, byte b5, byte b6, byte b7,
		byte b8, byte b9, byte bA, byte bB, byte bC, byte bD, byte bE, byte bF, byte x, byte y) {
	byte bytes[16] = {b0, b1, b2, b3, b4, b5, b6, b7, b8, b9, bA, bB, bC, bD, bE, bF};
	TotobotFace::showImage(bytes, sizeof bytes / sizeof *bytes, x, y);
}

void Totobot::showString(const char *str, int8_t x, int8_t y) {
	TotobotFace::showString(str, x, y);
}

void Totobot::setEyeEffect(byte eye, int effect) {
	TotobotFace::setEyeEffect(eye, effect);
}
